pkg/sentry/kernel/ipc/object.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package ipc defines functionality and utilities common to sysvipc mechanisms.
+package ipc
+
+import (
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/sentry/fs"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+	"gvisor.dev/gvisor/pkg/sync"
+)
+
+// Key is a user-provided identifier for IPC objects.
+type Key int32
+
+// ID is a kernel identifier for IPC objects.
+type ID int32
+
+// Object represents an abstract IPC object with fields common to all IPC
+// mechanisms.
+type Object struct {
+	// User namespace which owns the IPC namespace which owns the IPC object.
+	// Immutable.
+	UserNS *auth.UserNamespace
+
+	// ID is a kernel identifier for the IPC object. Immutable.
+	ID ID
+
+	// Key is a user-provided identifier for the IPC object. Immutable.
+	Key Key
+
+	// Creator is the user who created the IPC object. Immutable.
+	Creator fs.FileOwner
+
+	// Owner is the current owner of the IPC object.
+	Owner fs.FileOwner
+
+	// Perms is the access permissions the IPC object.
+	Perms fs.FilePermissions
+}
+
+// Mechanism represents a SysV mechanism that holds an IPC object. It can also
+// be looked at as a container for an ipc.Object, which is by definition a fully
+// functional SysV object.
+type Mechanism interface {
+	// Object returns a pointer to the mechanism's ipc.Object, and the mutex
+	// used to protect it.
+	Object() (*Object, *sync.Mutex)
```

Comment:
Instead of returning a mutex, define a `Lock` and `Unlock` method on the interface, and have the implementations call `mu.Lock`/`mu.Unlock`.

Reply:
Also, since you're grabbing locks on cross-package objects, you should probably document the lock ordering for relevant objects. For examples, see [shm](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/kernel/shm/shm.go;l=33?q=shm.go&ss=gvisor%2Fgvisor), or many of our filesystem implementations such as [goferfs](https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/fsimpl/gofer/gofer.go;l=18).

Reply:
Done